src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -814,61 +852,81 @@ private void CheckDisposed()
+            timer.Start();
+            _pingTimer.Reset();
+            _timeoutTimer.Reset();
 
-            try
-            {
-                Log.InvokingClosedEventHandler(_logger);
-                await closed.Invoke(closeException);
-            }
-            catch (Exception ex)
+            using (timer)
             {
-                Log.ErrorDuringClosedEvent(_logger, ex);
+                // await returns True until `timer.Stop()` is called in the `finally` block of `ReceiveLoop`
+                while (await timer)
+                {
+                    if (_timeoutTimer.Ready)
+                    {
+                        OnTimeout();
+                    }
+
+                    if (_pingTimer.Ready)
+                    {
+                        await PingServer();
+                    }
+                }
             }
         }
 
-        private void ResetTimeoutTimer(Timer timeoutTimer)
+        private void OnTimeout()
         {
-            if (timeoutTimer != null)
+            if (Debugger.IsAttached)
             {
-                Log.ResettingKeepAliveTimer(_logger);
-                timeoutTimer.Change(ServerTimeout, Timeout.InfiniteTimeSpan);
+                return;
             }
+
+            _connectionState.CloseException = new TimeoutException(
+                $"Server timeout ({_timeoutTimer.Interval.TotalMilliseconds:0.00}ms) elapsed without receiving a message from the server.");
+            _connectionState.Connection.Transport.Input.CancelPendingRead();
         }
 
-        private Timer StartTimeoutTimer(ConnectionState connectionState)
+        private async Task PingServer()
         {
-            // Check if we need keep-alive
-            Timer timeoutTimer = null;
+            if (_disposed || !_connectionLock.Wait(0))
+            {
+                // if we're disposed or can't get the lock, we don't need to ping
+                return;
+            }
+
+            // if we got here, we successfully have the lock
 
-            // We use '!== true' because it could be null, which we treat as false.
-            if (connectionState.Connection.Features.Get<IConnectionInherentKeepAliveFeature>()?.HasInherentKeepAlive != true)
+            try
             {
-                Log.StartingServerTimeoutTimer(_logger, ServerTimeout);
-                timeoutTimer = new Timer(
-                    state => OnTimeout((ConnectionState)state),
-                    connectionState,
-                    dueTime: ServerTimeout,
-                    period: Timeout.InfiniteTimeSpan);
+                if (_disposed || _connectionState == null || _connectionState.Stopping)
+                {
+                    return;
+                }
+
+                await SendHubMessage(PingMessage.Instance);
             }
-            else
+            finally
             {
-                Log.NotUsingServerTimeout(_logger);
+                ReleaseConnectionLock();
```

Comment:
There wont be a matching "Acquired connection lock" message currently

Reply:
Ooh, yeah good point :). Because you're not calling `WaitConnectionLock` and instead calling `Wait` directly on the semaphore we don't get the log message.